[Feature] Migrate hotkey uses to ngneat/hotkeys #73

EmmanuelBerkowicz opened the pull request:
# Description
feat: implement hotkey system using @ngneat/hotkeys

## Summary of Changes

- Install @ngneat/hotkeys package for centralized hotkey management
- Refactor game-play.component.ts to use lifecycle-based hotkey setup:
  - Import OnInit interface as type-only import
  - Import HotkeysService with ESLint disable (DI requirement vs type-only conflict)
  - Implement OnInit interface and move hotkey logic to ngOnInit()
  - Remove @HostListener decorators in favor of centralized hotkey service

Add hotkey mappings:
- 1: Toggle town menu
- 2: Toggle combat menu
- 3: Toggle inventory menu
- 4: Toggle heroes menu
- 5: Focus camera on player
- 6: Toggle options menu
- Space: Pause/resume game (feature already exist, changed method to new ng)
- Escape: Close all menus (feature already exist, changed method to new ng)

TODO: Add logout hotkey (q key)

Fixes #73 

[Feature] Migrate hotkey uses to ngneat/hotkeys

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [Y] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [Y] I have performed a self-review of my own code
- [Y] I have made corresponding changes to the documentation
- [Y] My changes generate no new warnings
- [Y] I have run tests (npm run test) that prove my fix is effective or that my feature works

EmmanuelBerkowicz commented:
Hey @seiyria ,
I've implemented the new hotkeys method you specified in #120, but I'm a bit stuck on how to implement the log out function.
Would appreciate some guidance on this.
Emmanuel.

seiyria commented:
Hey. Before I say anything else, I put this part in the template: `Fixes # (issue)` so you can replace it with the issue it fixes. Please actually fill this out by replacing (issue) with the number of the issue you're addressing. This is how github auto closes issues based on a PR merge.

Secondly, you should move all of this code to the navbar, probably. Then instead of copy-pasting all that code, you have much simpler functions to call (which already do this). To add the logic for quit, you need to add a `viewChild()` to get a ref to `leaveSwal` and then call `fire()` on the viewChild.

EmmanuelBerkowicz commented:
Thanks for the great feedback as usual @seiyria !

I've updated the PR and will keep this in mind next time :)

EmmanuelBerkowicz commented:
Hey @seiyria ,
I am having some issues with resolving the eslint error in the navbar.components.ts, line 45 @ViewChild, 'any'.
I wasn't able to figure out how to resolve this so I suppressed the ESLINT error for now.

If you figure out how to fix this can you please let me know?

Or if it's something obvious, can you please help point me in the right direction so I can fix it :)

seiyria commented:
I'm not sure, I've never seen those errors before.

EmmanuelBerkowicz commented:
@seiyria 
My understanding of why this was happening, was when you load the gamepage, the navbar component was created, which then registered the Hotkeys (onInit).

Then, when you nav to main menu, the navbar component is destroyed, but the hotkeys are still active in memory. Which means when you navigate back to game, you re create the navbar, but then because of the OnInit, you attempt to create new Hotkeys, which results in a duplicate which is the error message here.
So OnDestroy essentially destroys the hotkeys when the navbar is destroyed to remove this duplicate.

It also fixes the issue of the return to main menu pop up because the 'q' hotkeys was still listening even on main menu.

seiyria commented:
Right, that would be my guess too, but you handle the case of creation/destroy there, and you still get the errors? One option then is to hide, not destroy, the navbar where applicable (which is basically only the homepage).

EmmanuelBerkowicz commented:
@seiyria ,
No, I don't get the errors after I added the OnDestroy().
Sorry if I wasn't clear! I said that I added the OnDestroy() because of the errors noticed in the console and the action of the reoccurring return to menu interface.

seiyria commented:
Perfect! I haven't tested it yet, but then does it work as you intend? If so. I have some review and fixing to do tomorrow so I'll probably handle it then. Would have done so earlier if I knew it worked!
